File: Habitica/src/main/java/com/habitrpg/android/habitica/models/user/Stats.java
package com.habitrpg.android.habitica.models.user;

import android.content.Context;

import com.google.gson.annotations.SerializedName;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.models.HabitRpgClass;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;


public class Stats extends RealmObject {

    @PrimaryKey
    private String userId;

    User user;
    public Float con, str, per;
    @SerializedName("int")
    public Float _int;
    public Training training;
    public Buffs buffs;
    public Integer points, lvl;
    @SerializedName("class")
    public String habitClass;
    public Double gp, exp, mp, hp;
    private Integer toNextLevel, maxHealth, maxMP;


    public Integer getToNextLevel() {
        return toNextLevel != null ? toNextLevel : Integer.valueOf(0);
    }

    public void setToNextLevel(Integer toNextLevel) {
        if (toNextLevel != 0) {
            this.toNextLevel = toNextLevel;
        }
    }

    public Integer getMaxHealth() {
        return maxHealth != null ? maxHealth : Integer.valueOf(50);
    }

    public void setMaxHealth(Integer maxHealth) {
        this.maxHealth = maxHealth;
    }

    public Integer getMaxMP() {
        return maxMP != null ? maxMP : Integer.valueOf(0);
    }

    public void setMaxMP(Integer maxMP) {
        if (maxMP != 0) {
            this.maxMP = maxMP;
        }
    }



    public String getTranslatedClassName(Context context) {
        switch (habitClass) {
            case "healer":
                return context.getString(R.string.healer);
            case "rogue":
                return context.getString(R.string.rogue);
            case "warrior":
                return context.getString(R.string.warrior);
            case "wizard":
                return context.getString(R.string.mage);
            default:
                return context.getString(R.string.warrior);
        }
    }

    public void merge(Stats stats) {
        if (stats == null) {
            return;
        }
        this.con = stats.con != null ? stats.con : this.con;
        this.str = stats.str != null ? stats.str : this.str;
        this.per = stats.per != null ? stats.per : this.per;
        this._int = stats._int != null ? stats._int : this._int;
        this.training.merge(stats.training);
        this.buffs.merge(stats.buffs);
        this.points = stats.points != null ? stats.points : this.points;
        this.lvl = stats.lvl != null ? stats.lvl : this.lvl;
        this.habitClass = stats.habitClass != null ? stats.habitClass : this.habitClass;
        this.gp = stats.gp != null ? stats.gp : this.gp;
        this.exp = stats.exp != null ? stats.exp : this.exp;
        this.hp = stats.hp != null ? stats.hp : this.hp;
        this.mp = stats.mp != null ? stats.mp : this.mp;
        this.toNextLevel = stats.toNextLevel != null ? stats.toNextLevel : this.toNextLevel;
        this.maxHealth = stats.maxHealth != null ? stats.maxHealth : this.maxHealth;
        this.maxMP = stats.maxMP != null ? stats.maxMP : this.maxMP;
    }

    public String getHabitClass() {
        return habitClass;
    }

    public void setHabitClass(String habitClass) {
        this.habitClass = habitClass;
    }

    public Double getGp() {
        return gp;
    }

    public Double getExp() {
        return exp;
    }

    public Double getHp() {
        return hp;
    }

    public Double getMp() {
        return mp;
    }

    public Integer getLvl() {
        return lvl;
    }

    public void setExp(double exp) {
        this.exp = exp;
    }

    public void setHp(double hp) {
        this.hp = hp;
    }

    public void setGp(double gp) {
        this.gp = gp;
    }

    public void setMp(double mp) {
        this.mp = mp;
    }

    public void setLvl(Integer lvl) {
        this.lvl = lvl;
    }

    public Buffs getBuffs() {
        return buffs;
    }

    public Float getStr() {
        return str;
    }

    public Float get_int() {
        return _int;
    }

    public Float getCon() {
        return con;
    }

    public Float getPer() {
        return per;
    }

    public void setHabitClass(HabitRpgClass habitRpgClass) {
        habitClass = habitRpgClass.toString();
    }

    public String getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
        if (buffs != null && !buffs.isManaged()) {
            buffs.setUserId(userId);
        }
        if (training != null && !training.isManaged()) {
            training.setUserId(userId);
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/models/user/Stats.kt
package com.habitrpg.android.habitica.models.user

import android.content.Context
import com.google.gson.annotations.SerializedName
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.models.BaseObject
import com.habitrpg.shared.habitica.models.AvatarStats
import io.realm.RealmObject
import io.realm.annotations.RealmClass

@RealmClass(embedded = true)
open class Stats : RealmObject(), AvatarStats, BaseObject {
    @SerializedName("con")
    var constitution: Int? = null
    @SerializedName("str")
    var strength: Int? = null
    @SerializedName("per")
    var per: Int? = null
    @SerializedName("int")
    var intelligence: Int? = null
    var training: Training? = null
    override var buffs: Buffs? = null
    override var points: Int? = null
    override var lvl: Int? = null
    @SerializedName("class")
    override var habitClass: String? = null
    override var gp: Double? = null
    override var exp: Double? = null
    override var mp: Double? = null
    override var hp: Double? = null
    override var toNextLevel: Int? = null
        get() = if (field != null) field else 0
        set(value) {
            if (value != 0) {
                field = value
            }
        }
    override var maxHealth: Int? = null
        get() = if (field != null) field else 0
        set(value) {
            if (value != 0) {
                field = value
            }
        }
    override var maxMP: Int? = null
        get() = if (field != null) field else 0
        set(value) {
            if (value != 0) {
                field = value
            }
        }

    fun getTranslatedClassName(context: Context): String {
        return when (habitClass) {
            HEALER -> context.getString(R.string.healer)
            ROGUE -> context.getString(R.string.rogue)
            WARRIOR -> context.getString(R.string.warrior)
            MAGE -> context.getString(R.string.mage)
            else -> context.getString(R.string.warrior)
        }
    }

    fun merge(stats: Stats?) {
        if (stats == null) {
            return
        }
        this.constitution = if (stats.constitution != null) stats.constitution else this.constitution
        this.strength = if (stats.strength != null) stats.strength else this.strength
        this.per = if (stats.per != null) stats.per else this.per
        this.intelligence = if (stats.intelligence != null) stats.intelligence else this.intelligence
        this.training?.merge(stats.training)
        this.buffs?.merge(stats.buffs)
        this.points = if (stats.points != null) stats.points else this.points
        this.lvl = if (stats.lvl != null) stats.lvl else this.lvl
        this.habitClass = if (stats.habitClass != null) stats.habitClass else this.habitClass
        this.gp = if (stats.gp != null) stats.gp else this.gp
        this.exp = if (stats.exp != null) stats.exp else this.exp
        this.hp = if (stats.hp != null) stats.hp else this.hp
        this.mp = if (stats.mp != null) stats.mp else this.mp
        this.toNextLevel = if (stats.toNextLevel != null) stats.toNextLevel else this.toNextLevel
        this.maxHealth = if (stats.maxHealth != null) stats.maxHealth else this.maxHealth
        this.maxMP = if (stats.maxMP != null) stats.maxMP else this.maxMP
    }

    companion object {
        const val WARRIOR = "warrior"
        const val MAGE = "wizard"
        const val HEALER = "healer"
        const val ROGUE = "rogue"

        const val AUTO_ALLOCATE_FLAT = "flat"
        const val AUTO_ALLOCATE_CLASSBASED = "classbased"
        const val AUTO_ALLOCATE_TASKBASED = "taskbased"
    }
}
